mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -624,92 +624,67 @@ struct ThreadwiseWriteAllRewritePattern
-      sourceView.getType().cast<ShapedType>().getShape();
-  Type elemType = sourceView.getType().cast<MemRefType>().getElementType();
-  int64_t copyDPerThread = loadShape[1];
-  int64_t copyKPerThread = loadShape[0];
-  int64_t kpack = op.getKpack();
-
-  // We use kpackPerThread instead of kpack to cover edge cases where
-  // copyKPerThread is smaller than kpack
-  int64_t kpackPerThread = std::min(copyKPerThread, kpack);
-
-  Value rawLoadBuffer;
-  ArrayAttr loadBufferView;
-  bool needs64BitIdx;
-  std::tie(rawLoadBuffer, loadBufferView, needs64BitIdx) =
+  auto [rawLoadBuffer, loadBufferView, sourceNeeds64BitIdx] =
       untransform(b, sourceView);
-  assert(!needs64BitIdx && "Registers shouldn't need 64-bit indexing");
+  auto [rawStoreBuffer, storeBufferView, dstNeeds64BitIdx] =
+      untransform(b, destView);
+
+  assert(!sourceNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+  assert(!dstNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
   ArrayRef<int64_t> rawLoadBufferShape =
       rawLoadBuffer.getType().cast<ShapedType>().getShape();
-
-  Value rawStoreBuffer;
-  ArrayAttr storeBufferView;
-  std::tie(rawStoreBuffer, storeBufferView, needs64BitIdx) =
-      untransform(b, destView);
-  assert(!needs64BitIdx && "Registers shouldn't need 64-bit indexing");
   ArrayRef<int64_t> rawStoreBufferShape =
       rawStoreBuffer.getType().cast<ShapedType>().getShape();
-
+  if (rawLoadBufferShape.size() != 1)
+    return op.emitOpError("Raw load buffers have to be flat.");
+  if (rawStoreBufferShape.size() != 1)
+    return op.emitOpError("Raw store buffers have to be flat.");
   Value zero = b.createOrFold<arith::ConstantIndexOp>(loc, 0);
-  SmallVector<Value, 2> start(2, zero);
-  SmallVector<int64_t, 2> strides(2, 1);
-  int64_t vecLen = 1;
-  // The store buffer is a flattened < kouter x dPerThread x kpack >.
-  if (kpackPerThread == 1) {
-    // if kpack == 1, then we can do vectorized loads across d dimension from/to
-    // load/store buffer
-    vecLen = getMaxVectorizationForDatatype(loadBufferView, /*dim=*/1,
-                                            copyDPerThread, rawLoadBufferShape,
-                                            elemType);
-    vecLen = math_util::gcd(copyDPerThread, vecLen);
-    strides[1] = vecLen;
-  } else {
-    // if kpack > 1, then we are limited by vectorization in k dimension and it
-    // could be at most kpack.
-    vecLen = getMaxVectorizationForDatatype(loadBufferView, /*dim=*/0,
-                                            copyKPerThread, rawLoadBufferShape,
-                                            elemType);
-    vecLen = math_util::gcd(vecLen, kpackPerThread);
-    strides[0] = vecLen;
-  }
-  loadBufferView = collapseContiguousMerges(loadBufferView, rawLoadBufferShape);
-  storeBufferView =
-      collapseContiguousMerges(storeBufferView, rawStoreBufferShape);
+
+  auto storeBufferViewInverted = invertTransforms(b, loc, storeBufferView);
```

Comment:
Note: it looks like `invertTransforms` doesn't catch inversion failures properly, and this code should be testing for the return value of this being `nullptr` (you'll want to patch `invertTransforms` to propagate those up from `invertTransformMap`)